### PR TITLE
Update default opset to 14 in ORTModule

### PR DESCRIPTION
**Pull request by xadupre**
**Description**:

Update default opset to 14. The PR does the following kind of changes:

* update pytorch to 1.10 to support opset 14
* modifies the default opset to 14 in ortmodule
* updated expected value as when opset 14 was used (a couple of unit tests use hard coded expected values and not expected values computed with pytorch)
* updated torch version in dockers or added a parameter called `OPSET_VERSION` when parameter `PYTORCH_VERSION` was 
also an argument (possibles values are **1.9 + opset 12 or 1.10 + opset 14**). OPSET_VERSION is used to set ORTMODULE_ONNX_OPSET_VERSION.
* updated a couple file `requirements.txt` to pick the right versions of pytorch, torchvision, torchtext.
* added two unit tests to check onnxruntime gradient works for SoftmaxGrad and GatherGrad.

### CODE REVIEW

**Review by thiagocrepaldi — APPROVED** [review]

**Comment by baijumeswani** [comment]
Please update the file https://github.com/microsoft/onnxruntime/blob/2beb873c6bbf174e66f072d88ebd765bb94c6ff1/tools/ci_build/github/azure-pipelines/templates/orttraining-linux-gpu-ortmodule-test-ci-pipeline.yml#L36

this controls the torch version that is used in the `ORTModule` unit tests. If your changes depend on the torch 1.10 version, you would need to update this file.

**Comment by xadupre** [comment]
Not sure this is the only place. torch==1.9.1 is ok as it supports opset 14. torch==1.8.0 was not ok, the one I changed. Should I aligned all versions to 1.10?

**Comment by baijumeswani** [comment]
If it does not impact this PR directly, we can do it in another pull request. But we should update the torch version in the `ORTModule` unit tests..

File: tools/ci_build/github/linux/docker/Dockerfile.manylinux2014_training_cuda11_4 [hunk 1]
```diff
@@ -183,6 +183,7 @@ CMD ["/bin/bash"]
 #Build manylinux2014 docker image end
 ARG PYTHON_VERSION=3.6
 ARG TORCH_VERSION=1.9.0
+ARG OPSET_VERSION=12
```
Comment (baijumeswani) [inline tools/ci_build/github/linux/docker/Dockerfile.manylinux2014_training_cuda11_4:186]: why is this change required?

File: tools/ci_build/github/linux/docker/Dockerfile.manylinux2014_training_cuda11_4 [hunk 2]
```diff
@@ -200,3 +201,4 @@ RUN adduser --uid $BUILD_UID $BUILD_USER
 WORKDIR /home/$BUILD_USER
 USER $BUILD_USER
 ENV PATH /usr/local/gradle/bin:/usr/local/dotnet:$PATH
+ENV ORTMODULE_ONNX_OPSET_VERSION $OPSET_VERSION
```
Comment (baijumeswani) [inline tools/ci_build/github/linux/docker/Dockerfile.manylinux2014_training_cuda11_4:204]: why do we set the opset version as 12 in this docker file?

File: tools/ci_build/github/linux/docker/scripts/training/ortmodule/stage1/requirements_torch1.10.0_cu11.1.txt [hunk 3]
```diff
@@ -0,0 +1,6 @@
+--pre
```
Comment (baijumeswani) [inline tools/ci_build/github/linux/docker/scripts/training/ortmodule/stage1/requirements_torch1.10.0_cu11.1.txt:1]: why do we need 11.1 requirements file?

File: tools/ci_build/github/linux/docker/Dockerfile.manylinux2014_training_cuda11_3 [hunk 4]
```diff
@@ -183,6 +183,7 @@ CMD ["/bin/bash"]
 #Build manylinux2014 docker image end
 ARG PYTHON_VERSION=3.6
 ARG TORCH_VERSION=1.10.0
+ARG OPSET_VERSION=14
```
Comment (baijumeswani) [inline tools/ci_build/github/linux/docker/Dockerfile.manylinux2014_training_cuda11_3:186]: Why do we need to explicitly set this?

**Comment by baijumeswani** [comment]
Title says that the opset version is changed only for ORTModule, however the changes in the PR reflect opseet version change on other files (not impacting ORTModule) as well. Was that intended?

**Comment by xadupre** [comment]
Opset 14 needs torch>=1.10. Some dockers use a frozen version of torch which I updated, some others use a configurable version. The opset must depend on it. The number of changes is somehow proportional to the number of jobs runs on Azure DevOps.

File: tools/ci_build/github/linux/docker/Dockerfile.manylinux2014_training_cuda11_3 (hunk 4 above)
Comment (xadupre) [reply tools/ci_build/github/linux/docker/Dockerfile.manylinux2014_training_cuda11_3:186]: In that case, we don't need but I made this parameter explicit even if 14 is the default value. I did it to make it easier when we update default opset to 15.

**Comment by xadupre** [comment]
Some pipelines are explicitely checking onnxruntime + torch==1.9.0. This can only work with opset 12.

**Comment by baijumeswani** [comment]
> Opset 14 needs torch>=1.10. Some dockers use a frozen version of torch which I updated, some others use a configurable version. The opset must depend on it. The number of changes is somehow proportional to the number of jobs runs on Azure DevOps. I can reduce the number of impact files by selecting the opset based on Torch version (like the highest possible based on torch version).

In an [earlier comment](https://github.com/microsoft/onnxruntime/pull/9743#issuecomment-973096550) you mentioned that torch 1.9.1 was ok for opset 14. Not sure if opset 14 needs torch >= 1.10 or not. If it needs, may we please also update the torch version being used for the ORTModule unit tests.

File: orttraining/orttraining/python/training/_utils.py [hunk 5]
```diff
@@ -189,10 +189,14 @@ def decorate(func):
...
 
-    assert isinstance(file_path, str) and os.path.exists(file_path),\
-        "'file_path' must be a full path string with the python file to load"
-    assert module_name is None or isinstance(module_name, str) and module_name,\
-        "'module_name' must be a string with the python module name to load"
+    if not isinstance(file_path, str) or not os.path.exists(file_path):
+        raise AssertionError(
+            "'file_path' must be a full path string with the python file to load. "
+            "file_path=%r." % (file_path, ))
+    if not (module_name is not None or not isinstance(module_name, str) or not module_name):
```
Comment (baijumeswani) [inline orttraining/orttraining/python/training/_utils.py:196]: If `module_name` is `None`, we should probably not do the remainder of the checks in the `if` statement.
In the above if statement, if `module_name` is `None`, then we still go ahead and check the rest of the conditions.
I think the check should be:
`if module_name is not None and (not isinstance(module_name, str) or not module_name):`

File: orttraining/orttraining/test/python/onnxruntime_test_postprocess.py [hunk 6]
```diff
@@ -34,7 +34,7 @@ def get_onnx_model(self, model, model_desc, inputs, device,
                            device,
                            world_rank=0,
                            world_size=1,
-                           _opset_version=12,
+                           _opset_version=14,
```
Comment (baijumeswani) [inline orttraining/orttraining/test/python/onnxruntime_test_postprocess.py:37]: May we remove this line so that it uses the default (whatever is defined). Else, we might miss catching opset version regressions.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 7]
```diff
@@ -28,6 +28,7 @@
                                             LogLevel,
                                             _fallback,
                                             _graph_execution_manager)
+import onnxruntime.training.ortmodule as training_ortmodule
```
Comment (baijumeswani) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:31]: `training_ortmodule` gives the impression that there are multiple ortmodules. May we rename this to something like ortmodule_module or simply ortmodule?
import onnxruntime.training.ortmodule as ortmodule

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 8]
```diff
@@ -1560,6 +1565,7 @@ def run_step(model1, model2, model3, x1, x2):
         _test_helpers.assert_gradients_match_and_reset_gradient(ort_model1, pt_model1, atol=2e-6)
         _test_helpers.assert_gradients_match_and_reset_gradient(ort_model2, pt_model2)
         _test_helpers.assert_gradients_match_and_reset_gradient(ort_model3, pt_model3)
+    assert training_ortmodule.ONNX_OPSET_VERSION == DEFAULT_OPSET
```
Comment (baijumeswani) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:1568]: Instead of adding this line to many (unrelated) tests, may we just add a single unit test to check the default onnx opset version?

File: orttraining/orttraining/test/python/orttraining_test_onnx_ops_ortmodule.py [hunk 9]
```diff
@@ -0,0 +1,154 @@
+"""
```
Comment (baijumeswani) [inline orttraining/orttraining/test/python/orttraining_test_onnx_ops_ortmodule.py:1]: Is this test being invoked from any file?

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py [hunk 10]
```diff
@@ -1499,25 +1515,44 @@ def _adam_max_norm_clip_data():
...
+            (0, 'cuda', 0.1, 1, 12, {
+                12: [10.592951, 10.068722, 9.620503, 9.247791, 8.883972,
+                     8.582286, 8.285027, 8.068308, 7.803638, 7.492318, 7.240352, 7.018665],
+                14: [10.584141, 10.068845, 9.583107, 9.193537, 8.882966, 8.538839,
+                     8.315872,  8.066408, 7.778978, 7.529708, 7.265849, 7.035439]}),
+            (42, 'cuda', 1.0, 1, 12, {
+                12: [10.647908, 10.144501, 9.672352, 9.306980, 8.956026,
+                     8.602655, 8.351079, 8.088144, 7.867220, 7.564082, 7.289846, 7.073726],
+                14: [10.697515, 10.229034, 9.765422, 9.428294, 9.080612, 8.715208,
+                     8.459574, 8.169073, 7.940211, 7.654147, 7.390446, 7.166227]}),
```
Comment (baijumeswani) [inline orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py:1532]: Are you missing `(42, 'cuda', 0.1, 1, 12...` here?

File: orttraining/orttraining/test/python/orttraining_test_utils.py [hunk 11]
```diff
@@ -123,7 +123,7 @@ def run_test(model, model_desc, device, args, gradient_accumulation_steps, fp16,
             allreduce_post_accumulation=allreduce_post_accumulation,
             get_lr_this_step=get_lr_this_step if use_internal_get_lr_this_step else None,
             loss_scaler=loss_scaler if use_internal_loss_scaler else None,
-            _opset_version=12,
+            _opset_version=14,
```
Comment (baijumeswani) [inline orttraining/orttraining/test/python/orttraining_test_utils.py:126]: shall we let this be the default setting?

File: tools/ci_build/github/azure-pipelines/orttraining-pai-ci-pipeline.yml [hunk 12]
```diff
@@ -65,6 +65,7 @@ jobs:
   - script: |-
       cd ./build/RelWithDebInfo
       export PYTHONPATH=$PWD
+      export ORTMODULE_ONNX_OPSET_VERSION=12
```
Comment (baijumeswani) [inline tools/ci_build/github/azure-pipelines/orttraining-pai-ci-pipeline.yml:68]: Are we not able to update the torch versin for this pipeline?

**Comment by baijumeswani** [comment]
> In an [earlier comment](https://github.com/microsoft/onnxruntime/pull/9743#issuecomment-973096550) you mentioned that torch 1.9.1 was ok for opset 14. Not sure if opset 14 needs torch >= 1.10 or not. If it needs, may we please also update the torch version being used for the ORTModule unit tests.

Never mind. Just saw that the pull request already updated the ortmodule tests to run with torch 1.10. Thanks.

File: orttraining/orttraining/core/optimizer/megatron_transformer.cc [hunk 13]
```diff
@@ -33,19 +33,19 @@ const std::initializer_list<ONNX_NAMESPACE::OperatorSetVersion> opset_v1_13 = {1
...
 const OpInfo split_info = OpInfo("Split", opset_v2_11_13, kOnnxDomainAlias, 3);
 const OpInfo reshape_info = OpInfo("Reshape", opset_v5_13);
 const OpInfo transpose_info = OpInfo("Transpose", opset_v1_13);
 const OpInfo matmul_info = OpInfo("MatMul", opset_v9_13);
-const OpInfo div_info = OpInfo("Div", opset_v7_13);
-const OpInfo mul_info = OpInfo("Mul", opset_v1_6_7_13);
-const OpInfo sub_info = OpInfo("Sub", opset_v7_13);
+const OpInfo div_info = OpInfo("Div", opset_v7_13_14);
+const OpInfo mul_info = OpInfo("Mul", opset_v1_6_7_13_14);
+const OpInfo sub_info = OpInfo("Sub", opset_v7_13_14);
```
[the remark this answers is not in the document]
Comment (xadupre) [reply orttraining/orttraining/core/optimizer/megatron_transformer.cc:48]: I did not check if the unit tests were failing without these changes. I assumed the opsets mentioned below were needed as operators Add, Mul, Sub, Sub were updated in opset 14.

File: orttraining/orttraining/python/training/_utils.py (hunk 5 above)
Comment (xadupre) [reply orttraining/orttraining/python/training/_utils.py:196]: done

File: orttraining/orttraining/test/python/onnxruntime_test_postprocess.py (hunk 6 above)
Comment (xadupre) [reply orttraining/orttraining/test/python/onnxruntime_test_postprocess.py:37]: The default is 14 for ORTModule but this is an explicit parameter given to ORTTrainer.

File: orttraining/orttraining/test/python/orttraining_test_onnx_ops_ortmodule.py (hunk 9 above)
Comment (xadupre) [reply orttraining/orttraining/test/python/orttraining_test_onnx_ops_ortmodule.py:1]: It is a standalone test. I checks the correctness for SoftmaxGrad and GatherGrad. There was no python unit test checking explicitely GatherGrad. I added these to check explicity the gradient of gradient operator on CPU and GPU on very small networks.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 7 above)
Comment (xadupre) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:31]: I chose `training_ortmodule` because `ortmodule` is used a variable in one unit test. I can change.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 8 above)
Comment (xadupre) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:1568]: I'll change. I did it that way to avoid creating a new test and be faster.

File: orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py (hunk 10 above)
Comment (xadupre) [reply orttraining/orttraining/test/python/orttraining_test_orttrainer_frontend.py:1532]: Thanks, I missed it.

File: orttraining/orttraining/test/python/orttraining_test_utils.py (hunk 11 above)
Comment (xadupre) [reply orttraining/orttraining/test/python/orttraining_test_utils.py:126]: The value was explicitely set even though the default opset was 12. I did not change that.

File: tools/ci_build/github/linux/docker/scripts/training/ortmodule/stage1/requirements_torch1.10.0_cu11.1.txt (hunk 3 above)
Comment (xadupre) [reply tools/ci_build/github/linux/docker/scripts/training/ortmodule/stage1/requirements_torch1.10.0_cu11.1.txt:1]: One job is using CUDA 11.1.

File: tools/ci_build/github/azure-pipelines/orttraining-pai-ci-pipeline.yml (hunk 12 above)
Comment (xadupre) [reply tools/ci_build/github/azure-pipelines/orttraining-pai-ci-pipeline.yml:68]: I guess we could. don't know which value to choose for this parameter: https://github.com/microsoft/onnxruntime/blob/master/tools/ci_build/github/pai/rocm-ci-pipeline-env.Dockerfile#L1.

File: orttraining/orttraining/test/python/orttraining_test_onnx_ops_ortmodule.py (hunk 9 above)
Comment (baijumeswani) [reply orttraining/orttraining/test/python/orttraining_test_onnx_ops_ortmodule.py:1]: Do we not want to invoke it in the CI pipelines?

Comment (xadupre) [reply orttraining/orttraining/test/python/orttraining_test_onnx_ops_ortmodule.py:1]: I assumed it was automatically added. I added the file to the list of test file to run.

**Review by baijumeswani — APPROVED** [review]

**Comment by ganik** [comment]
@xadupre  I am getting this error after this upgrade to opset 14 on previously successfully run tests in one of my converrsion projects. If I downgrade to opset 12 this error disappears ....

E           RuntimeError: /onnxruntime_src/orttraining/orttraining/python/orttraining_pybind_state.cc:744 onnxruntime::python::addObjectMethodsForTraining(pybind11::module&, onnxruntime::python::ExecutionProviderRegistrationFn)::<lambda(onnxruntime::training::OrtModuleGraphBuilder*)> [ONNXRuntimeError] : 10 : INVALID_GRAPH : This is an invalid model. In Node, ("ConcatTraining_token_19_Grad/Split_0", Split, "", -1) : ("expert_output_grad": tensor(float),) -> ("273_grad": tensor(float),) , Error Unrecognized attribute: split for operator Split

**Comment by xadupre** [comment]
@ganik, operator Split changes in opset 13, attribute split became an input in opset 13. I think one operator Split is described following opset 13 in a graph following opset 14. How did you get that error?
